### ISSUE TITLE
Cannot fetch from icr.io

### ISSUE THREAD

**Issue by bogdanbrindusan** [GoogleContainerTools/jib#3700]
I've been using JIB successfully for a long time now, however I'm trying to make use of IBM Semeru Java runtime.
The certified builds are not available in Docker Hub, only in icr.io - IBM's Container registry.

The integration between JIB & IBM's Container registry doesn't work.

**Environment**:

- *Jib version:* 3.2.1
- *Build tool:* Gradle 7.3.3
- *OS:* MacOS 12.4

**Description of the issue**:
Setting JIB to use a base image from icr.ro ends up throwing the following error:
```
* What went wrong:
Execution failed for task ':jibDockerBuild'.
> com.google.cloud.tools.jib.plugins.common.BuildStepsExecutionException: Tried to pull image manifest for icr.io/appcafe/ibm-semeru-runtimes:certified-17-jdk-ubi-minimal but failed because: OCI index found, but accept header does not support OCI indexes
```
Doing a `docker pull` on the image works though:
```
docker pull icr.io/appcafe/ibm-semeru-runtimes:certified-17-jdk-ubi-minimal
certified-17-jdk-ubi-minimal: Pulling from appcafe/ibm-semeru-runtimes
54e56e6f8572: Pull complete
4f8ddd7f5a75: Pull complete
e5393f9ab5da: Pull complete
b33516b9ff85: Pull complete
64aeed9e771e: Pull complete
Digest: sha256:5849ede694ee8791cc4369afed2e1fd4b2b36b971b0792a4cc9069e042c25a31
Status: Downloaded newer image for icr.io/appcafe/ibm-semeru-runtimes:certified-17-jdk-ubi-minimal
```

**Expected behavior**:
Docker image is built successfully using the base image from icr.io

**`jib-gradle-plugin` Configuration**: 
```groovy
jib {
    from {
        image = 'icr.io/appcafe/ibm-semeru-runtimes:certified-17-jdk-ubi-minimal'
    }
    to {
        image = "personal/${rootProject.name}:" + version
    }
    extraDirectories {
        paths {
            path {
                from = file('src/docker/')
                into = file('/app/')
            }
            path {
                from = file("$buildDir/newrelic/")
                into = file('/app/')
            }
        }
    }
    container {
        jvmFlags = ['-javaagent:/app/newrelic.jar']
        creationTime = 'USE_CURRENT_TIMESTAMP'
    }
}
```

**Log output**: 
```
* What went wrong:
Execution failed for task ':jibDockerBuild'.
> com.google.cloud.tools.jib.plugins.common.BuildStepsExecutionException: Tried to pull image manifest for icr.io/appcafe/ibm-semeru-runtimes:certified-17-jdk-ubi-minimal but failed because: OCI index found, but accept header does not support OCI indexes
```

**Comment by chanseokoh** [comment]
This is a dup of #2749.

The reference `icr.io/appcafe/ibm-semeru-runtimes:certified-17-jdk-ubi-minimal(:latest)` is an OCI index, which is a list of pointers to multiple images of different platforms. In the [FAQ](https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#how-do-i-specify-a-platform-in-the-manifest-list-or-oci-index-of-a-base-image), we say

> OCI image indices are not supported (as opposed to Docker manifest lists).

(An OCI index is the OCI counterpart of the Docker manifest list format. Actually both formats are mostly same in practice.)

For now, the only workaround is to explicitly specify the SHA digest of the image for the platform you want to use (e.g., `icr.io/appcafe/ibm-semeru-runtimes:certified-17-jdk-ubi-minimal:latest@sha256:xxxxxxxxx...`).

**Comment by bogdanbrindusan** [comment]
@chanseokoh I'm trying to use your tip to get it working but I'm unsuccessful. 
It doesn't work putting `latest` as you suggested, as the tag is already there.
I've tried setting it like this:
```
image = 'icr.io/appcafe/ibm-semeru-runtimes:certified-17-jdk-ubi-minimal@sha256:5849ede694ee8791cc4369afed2e1fd4b2b36b971b0792a4cc9069e042c25a31'
```
But I still get same error:
```
* What went wrong:
Execution failed for task ':jibDockerBuild'.
> com.google.cloud.tools.jib.plugins.common.BuildStepsExecutionException: Tried to pull image manifest for icr.io/appcafe/ibm-semeru-runtimes:sha256:5849ede694ee8791cc4369afed2e1fd4b2b36b971b0792a4cc9069e042c25a31 but failed because: OCI index found, but accept header does not support OCI indexes
```
Any other suggestions are very welcome.

**Comment by chanseokoh** [comment]
Your SHA is still pointing to an OCI index. You should point it to an individual image.

Using [`crane`](https://github.com/google/go-containerregistry/blob/main/cmd/crane/README.md) (normally I use [`docker manifest inspect`](https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#how-do-i-specify-a-platform-in-the-manifest-list-or-oci-index-of-a-base-image), but for some reason, `icr.io` doesn't work with it), I see the index has the following image pointers:

```
$ crane manifest icr.io/appcafe/ibm-semeru-runtimes:certified-17-jdk-ubi-minimal@sha256:5849ede694ee8791cc4369afed2e1fd4b2b36b971b0792a4cc9069e042c25a31 | jq
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "manifests": [
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:c821b606ab3afeaba9ab498d6703ae90aed398947e6ec14eece63d15da90cf35",
      "size": 1255,
      "platform": {
        "architecture": "s390x",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "digest": "sha256:3a832b84c9a96134dd439c7ba60c309e5857a8964ff34005cfc7559101235f58",
      "size": 1373,
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:3acd2e5f2b7ba66d05feca81d6ae6b3f4450aed0bb89bb34689e428f6bfdf8ab",
      "size": 1255,
      "platform": {
        "architecture": "ppc64le",
        "os": "linux"
      }
    }
  ]
}
```

So, for example, if you want to pull a linux/amd64 image, the digest is `sha256:3a832b84c9a96134dd439c7ba60c309e5857a8964ff34005cfc7559101235f58`.

> It doesn't work putting `latest` as you suggested, as the tag is already there.

I think this is the issue of `icr.io`. In other registries, `:latest` in this case only serves as a comment.

---

It's funny that the OCI index mixes Docker format (`application/vnd.docker.distribution.manifest.v2+json`) and OCI format (`application/vnd.oci.image.manifest.v1+json`), but that shouldn't be a problem.

**Comment by chanseokoh** [comment]
> I'll have to manually update to newer versions when they become available

Yes. But we generally recommend explicitly [specifying a digest for strong reproducibility](https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#where-is-bash).

**Comment by bogdanbrindusan** [comment]
@chanseokoh we moved away from that as we have a fast pace of releasing and for us it's more important to be on latest JDK versions as automatically as possible. This reduces maintenance work which allows us to spend time on other things.

But the bottomline is that this is an issue and I'm looking forward seeing it fixed.
I'll have a look if I can contribute myself.

**Comment by chanseokoh** [comment]
> I'll have a look if I can contribute myself.

I'll document what I can remember.

The [OCI format](https://github.com/opencontainers/image-spec/blob/main/image-index.md#example-image-index) and the [Docker format](https://docs.docker.com/registry/spec/manifest-v2-2/#example-manifest-list) for a "manifest list" (or "OCI index") look very much the same for our purposes, so it's not really that difficult to add support for OCI indexes. The [`OCIIndexTemplate`](https://github.com/GoogleContainerTools/jib/blob/master/jib-core/src/main/java/com/google/cloud/tools/jib/image/json/OciIndexTemplate.java) class already exists, which is the OCI counterpart of the Docker format one [`V22ManifestListTemplate`](https://github.com/GoogleContainerTools/jib/blob/master/jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestTemplate.java). Both template classes inherit `ManifestListTemplate`.

You'll need to update [`AbstractManifestPuller.getAccept()`](https://github.com/GoogleContainerTools/jib/blob/78f069b6dcb2ce1edbb43dc909006d7bf905fe85/jib-core/src/main/java/com/google/cloud/tools/jib/registry/AbstractManifestPuller.java#L70-L89) and declare that it accepts the OCI index. The [detection and conversion logic](https://github.com/GoogleContainerTools/jib/blob/78f069b6dcb2ce1edbb43dc909006d7bf905fe85/jib-core/src/main/java/com/google/cloud/tools/jib/registry/AbstractManifestPuller.java#L165-L176) should also be updated.

Next, implement iterative over the OCI index in [`PullBaseImageStep`](https://github.com/GoogleContainerTools/jib/blob/78f069b6dcb2ce1edbb43dc909006d7bf905fe85/jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java#L317-L349). That part assumes `manifestTemplate` is `V22ManifestListTemplate`).
